### PR TITLE
#1550 #1706 Addressing the QoS options ExceptionsAllowedBeforeBreaking issue

### LINKED ISSUE #1550: Circuit breaker not handling exception at end points even when ExceptionsAllowedBeforeBreaking is set.
## Expected Behavior / New Feature

As per the documentation of QoS functionality of Ocelot, it uses Polly to achieve Circuit Breaker among the other resilience features. From what I could understand, the following configuration should have triggered the circuit breaker after an endpoint raised exception more than 2 times.

```json
"QoSOptions": {
   "ExceptionsAllowedBeforeBreaking": 2,
   "DurationOfBreak": 10000,
   "TimeoutValue": 2000
}
```

## Actual Behavior / Motivation for New Feature

However, it looked like the exception raised by the end points isn't handled by the circuit breaker. Please note that the Timeout works perfectly and triggers the circuit breaker as expected. However the case wasn't same with the _ExceptionsAllowedBeforeBreaking_.

## Steps to Reproduce the Problem

  1. Create a route in the gateway with the above QoS configuration
  2. The specified Endpoint should raise an exception 
  3. Raise the request for end point via the gateway.

The code sample in the following link demonstrates the issue
[Demo Code](https://github.com/anuviswan/LearningPoint/tree/master/Microservices/AuthenticationDemo)

Similar problem is mentioned in the following Stackoverflow question too
- [Ocelot QoS doesn't break circuit when n number of exception arises](https://stackoverflow.com/questions/56851572/ocelot-qos-doesnt-break-circuit-when-n-number-of-exception-arises)

## Specifications
  - Ocelot: 17.0.1
  - Ocelot.Provider.Polly: 17.0.1

### PR DESCRIPTION
**Pull request by ggnaegi**
## Fixes #1550 #1706 
- #1550 
- #1706 

## Proposed Changes
- We shouldn't create a Circuit Breaker per route / request, only maintain one Circuit Breaker per route
- The Circuit Breaker is Expecting an Exception to kick in, a response with status 500 is not an Exception. If status 500, then throwing.

### CODE REVIEW

**Review by ks1990cn — APPROVED** [review]

**Review by ks1990cn — APPROVED** [review]

**Review by ks1990cn — APPROVED** [review]

**Review by ks1990cn — APPROVED** [review]

**Comment by ggnaegi** [comment]
@raman-m @RaynaldM This PR is ready for review. 

I have added several unit tests:
- Circuit breaker should build
- it contains only one policy if ExceptionsAllowedBeforeBreaking = 0
- it contains two policies of type ```AsyncCircuitBreakerPolicy<HttpResponseMessage>``` and ```AsyncTimeoutPolicy<HttpResponseMessage>``` if ExceptionsAllowedBeforeBreaking > 0
- should return same circuit breaker for given route
- should return different circuit breakers for two different routes
- should throw exception ```BrokenCircuitException<HttpResponseMessage>```` on third faulty attempt if ExceptionsAllowedBeforeBreaking = 2
- should no throw if status code is OK and not within 500...508 range
- should throw but after delay allow one new request
- should throw but atfer delay allow one faulty request and then throw again
- should throw but atfer delay allow one correct response and put counter back to zero
acceptance tests:
- should open circuit breaker after two exceptions

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
Gui, would you like to include this fix in October release?

**Comment by ggnaegi** [comment]
> Gui, would you like to include this fix in October release?

I think so yes.

**Comment by raman-m** [comment]
Gui,
As you see, I've rebased feature branch... And I'll review tomorrow...
Happy presentation tomorrow! 😉

File: Ocelot.sln.DotSettings
Comment (raman-m) [inline Ocelot.sln.DotSettings:2]: New line before EOF is expected

File: src/Ocelot.Provider.Polly/CircuitBreaker.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/CircuitBreaker.cs:14]: Are you sure that 1st array item will be `CircuitBreakerAsyncPolicy` matching the type?
Maybe you need a predicate by type here...

File: src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs:29]: Not an issue. But could we rename it removing suffix `Delegate` ?
According to Microsoft .NET conventions, [EventHandler](https://learn.microsoft.com/en-us/dotnet/api/system.eventhandler?view=net-7.0) or other `Handler` types are delegates.
So, there is no need to add this suffix, it is redundant.

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:48]: For what purpose do you re-throw? To be caught by Exception-to-Error logic?

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:50]: This is the last exception type we process. What about the rest of `Exception` objects?
Are they caught by Exception-to-Error logic?

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyQoSProvider.cs:1]: Please, show me the current Usings settings in Visual Studio which put system namespaces on top of all.
We need to use simple Remove and Sort refactoring...

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (raman-m) [inline src/Ocelot.Provider.Polly/PollyQoSProvider.cs:63]: Cannot understand that: `_serverErrorCodes` ...
Why do we handle all these HTTP statuses?
Is that some kind of best practice for Polly?

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (raman-m) [inline src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]:
```suggestion
    public delegate DelegatingHandler QosDelegatingHandler(DownstreamRoute route, IHttpContextAccessor contextAccessor, IOcelotLoggerFactory loggerFactory);
```
So, remove "Delegate" suffix

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/PollyQoSTests.cs:22]: Please, setup configuration being returned by private helper.
A lot of copy-pasted code!

File: test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs:191]: Strange redundant empty methods because of the `IAsyncPolicy` interface implementation.
Could we use `Mock<IAsyncPolicy>` or `Mock<AsyncPolicy<TResult>>` instead of `FakeAsyncPolicy<TResult>`?
Is it possible to use mocks to test?

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/PollyQoSTests.cs:194]: Did you fix this #1706 unstable test? 😍 
Last fail which known for me is build [2454](https://circleci.com/gh/ThreeMammals/Ocelot/2454)

**Review by raman-m — CHANGES_REQUESTED** [review]
Not bad! Good job! 👍 
Let's do some final improvements...

File: src/Ocelot.Provider.Polly/CircuitBreaker.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/CircuitBreaker.cs:14]: We could just call it OcelotAsyncPolicy? The policies are wrapped together here.

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:48]: That's the original logic, I didn't want to mess with it.

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs:50]: Same as above...

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:1]: Sorry, I keep forgetting your settings

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:63]: That was an assumption. Http Status is returned from SendAsync, polly will react on certain types. I thought 500...508 is great.

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: Ok

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:22]: Ok, factories.

File: test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs
Comment (ggnaegi) [reply test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs:191]: Yeah, ok, that's the original code, let me check that

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:194]: I hope so, I simplified the logic yes.

File: Ocelot.sln.DotSettings
Comment (ggnaegi) [reply Ocelot.sln.DotSettings:2]: It shouldn't be part of code, it's resharper related, let's put that in the ignore list

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:22]: @raman-m but that's not my code...

Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:22]: Let's keep it like that for the minute?

File: src/Ocelot.Provider.Polly/CircuitBreaker.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/CircuitBreaker.cs]: @raman-m Ok, since the timeout is part of the policies, it can't be named CircuitBreaker. I changed the names to be more precise: public PollyPolicyWrapper and then AsyncPollyPolicy for the wrapped policies.

File: Ocelot.sln.DotSettings
Comment (raman-m) [reply Ocelot.sln.DotSettings]: 😮  Do you use ReSharper to write code for this project?
I **do not** recommend to use JetBrains tools for this project. I believe a developer should use Visual Studio IDE only. The project has standard settings styling, conventions, merge tools etc. for Visual Studio only.
That is why I see so strange diff highlighted by GitHub diff comparer in PRs... Keep in mind, that we have special line-ending settings!

If you use ReSharper IDE, please do not add any ReSharper files to the repo!

File: src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs:29]: Not changed! The [link](https://github.com/ThreeMammals/Ocelot/pull/1753/files#diff-586799ba26fd065bccd320cb27577637fa83b08a9ec758a28c792294115086b0R7)
It returns `DelegatingHandler`!

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs]: OK. I hate this Exception-to-Error mapping too. Bad design! We have a lot of problems because of this feature...
Global UnhandledException handler should catch exceptions for logging purposes. Only for rare cases we could catch exception to stabilize the behavior...

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs]: I mean, we need to ignore other exceptions here, right?

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs]: Not mine only 🤣  for whole community!
I don't see any benefits at all to shuffle usings lines. Keeping this simplest is better. Remove and Sort refactoring is ideal candidate to keep usings good.

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:63]: If internal Polly implementations use these status codes (or overrides them) then it is OK for me to process them.

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (raman-m) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: Not OK! 😛 Where is the change? The proof [here](https://github.com/ThreeMammals/Ocelot/pull/1753/files#diff-586799ba26fd065bccd320cb27577637fa83b08a9ec758a28c792294115086b0R7)

Why not to press the magic button "Commit suggestion"? 😉 
The rest of work is renaming all references.
OR
You can rename manually by pressing F2 button onto the class definition. This is the best option.

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs]: Let me help you...

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: I was wondering if we could just remove it. Why this abstraction?

File: src/Ocelot.Provider.Polly/PollyQoSProvider.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyQoSProvider.cs:63]: Yes it reacts on the status codes you choose, or other specs from HttpResponseMessage class... With that we can avoid the overhead of throwing an exception.

File: Ocelot.sln.DotSettings
Comment (ggnaegi) [reply Ocelot.sln.DotSettings]: @raman-m As I wrote once, I'm on several projects, sometimes I forget... That's why I'm happy to work with quality gates.

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (ggnaegi) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs]: Mmmh, I have refactored it already. But it wasn't my code again..

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (ggnaegi) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: Is it of any use?

Comment (ggnaegi) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: I mean, do we have other qos implementations? Why bother?

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (ggnaegi) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs]: Imho, it should be all the exceptions should be catched or none, but now it's, sometimes yes, sometimes no, it's not consistent.

**Comment by ggnaegi** [comment]
@raman-m are you sure that it's not a bit over-engineered?

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs]: Just commit this https://github.com/ThreeMammals/Ocelot/pull/1753/commits/50f9b7a928da0c84e19eacf77dd258c0cc42a33f
I'm sorry, if we did this in parallel.

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [reply test/Ocelot.AcceptanceTests/PollyQoSTests.cs:109]: Thanks!

**Comment by raman-m** [comment]
> @ggnaegi commented [on Nov 3](#issuecomment-1792422894)
 a bit over-engineered?

Why not if I have some time for this...
I agree that refactoring of tests doesn't provide much benefit.
🆗 I won't you bother on tests... Let's focus on real code...

File: src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs
Comment (raman-m) [reply src/Ocelot/Requester/QosDelegatingHandlerDelegate.cs:7]: A developer's joke... `DelegatingHandlerDelegateHandlerDelegateFactory` 😄 
I saw such classes written by Tom.

File: src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs
Comment (raman-m) [reply src/Ocelot.Provider.Polly/PollyCircuitBreakingDelegatingHandler.cs]: OK Let's ignore for now...

File: test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs
Comment (raman-m) [reply test/Ocelot.UnitTests/Polly/PollyCircuitBreakingDelegatingHandlerTests.cs]: If refactoring requires a lot of effort, then it is better to ignore it.

File: .gitignore
Comment (raman-m) [inline .gitignore:258]: Note 1

File: test/Ocelot.UnitTests/Configuration/Validation/FileConfigurationFluentValidatorTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Configuration/Validation/FileConfigurationFluentValidatorTests.cs:17]: Note 2 Order

File: test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs:181]: Note 3

File: test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs:371]: Note 3 end

**Review by raman-m — APPROVED** [review]
Great job, Gui! 👍 
You've got approval! 😄 
If you don't mind, I will add a couple of commits...

**Comment by raman-m** [comment]
@ggnaegi Let me know you're 🆗 to merge plz?
I'm going to add a couple of tiny fixes...
So, it seems, this PR will be merged first than #1745  ... And you will need to apply new logger factories for all new logging operators being added by this PR.

File: .gitignore
Comment (raman-m) [reply .gitignore:258]: @ggnaegi 
Could you send me a full list of ReSharper artifacts being added as standard settings by the tool to .NET solutions please?

I'll add them to ignore list...
In future, if someone will convert, or create std setup of ReSharper tools then we readd them and remove from **.gitignore** file.
But for now, I want to make sure that contributors don't change line-endings at all. This is critical settings for code review.

Comment (raman-m) [reply .gitignore:258]: What's your vision to fix the problem with line-endings in ReSharper?

**Comment by raman-m** [comment]
Regarding the [Open_circuit_should_not_effect_different_route()](/ggnaegi/Ocelot/blob/bug/1550_circuit_breaker_not_handling_exception_at_end_points/test/Ocelot.AcceptanceTests/PollyQoSTests.cs#L109) test... Here is my testing results of 1000 runs for the test;

image

It is definitely stable now. But I don't know how to get CircleCI environment tests... Seems we have to watch for this test on CircleCI in the latest builds after merging PR...

Anyway, I would say, It should be stable now, and #1706 will **not be** reproducible after merging this PR.
So, we could attach #1706 to this PR.
Could we, Gui?

File: test/Ocelot.UnitTests/Configuration/Validation/FileConfigurationFluentValidatorTests.cs
Comment (raman-m) [reply test/Ocelot.UnitTests/Configuration/Validation/FileConfigurationFluentValidatorTests.cs]: Fixed in https://github.com/ThreeMammals/Ocelot/pull/1753/commits/6504e3ef6d3c3a1d70a57bc9100f8f7397fa427a

File: test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs
Comment (raman-m) [reply test/Ocelot.UnitTests/Polly/PollyPoliciesDelegatingHandlerTests.cs]: Fixed in https://github.com/ThreeMammals/Ocelot/pull/1753/commits/4e5f1fa72338f77de206fad812b76a666b3bc2cd ✔️

**Comment by ggnaegi** [comment]
Great!

**Pull request merged**
